server/lib/models/maps.js

```diff
@@ -301,6 +302,40 @@ export async function updateMap(mapId, favoritedBy, name, description, visibilit
   }
 }
 
+// transactionFn is a function() => Promise, which will be awaited inside the DB transaction. If it
+// is rejected, the transaction will be rolled back.
+export async function deleteMap(mapId, transactionFn) {
+  return transact(async client => {
+    let query = `
+      WITH fav AS (
+        DELETE FROM favorited_maps
+        WHERE map_id = $1
+      )
+      DELETE FROM uploaded_maps
+      WHERE id = $1;
+    `
+    const params = [mapId]
+
+    await client.query(query, params)
+
+    // Delete all maps that no one is referencing anymore, if any
+    query = `
+      DELETE FROM maps AS m
+      WHERE m.hash NOT IN (
+        SELECT map_hash
+        FROM uploaded_maps
+      )
+      RETURNING *;
+    `
```

Comment:
Would having the first deletion query return map hash and limiting this query to only operate on map with that hash be better?

(To be clear, I have no idea how this currently performs relatively or absolutely, or if we have any reason to believe that this query may need to delete more than one map ever)

Reply:
Hmm, I'm having trouble imagining how that would work?
We need the check if the hash of the deleted map exists at all in the `uploaded_maps` table before deleting an entry in the `maps` table.

Granted, this doesn't look like the most optimal way to do it, but I'm really failing to see how else it could be done at the moment.

Reply:
Oh nvm, I see what you mean now (I think :d).

Reply:
Nvm again, I don't think I know what you mean; or at least not enough to make it work ^^

Reply:
Yeah, I don't have too clear concept of it to give any concrete idea how to do it. Current way is fine to me